### ISSUE TITLE
Sloped terrain destroys detailed mapping

### ISSUE THREAD

**Issue by ivanbranco** [StrandedKitty/streets-gl#58]
As @endim8 wrote in 48:

> i think this is because each building part gets its own offset from the sloped terrain, as opposed to getting one offset from the main building area and using it for all the contained building parts. But it's best to make a new issue for this :)

Examples: https://streets.gl/#51.51475,-0.09843,20.50,354.50,415.75 ![image](https://user-images.githubusercontent.com/26939824/236321953-a6120055-9b09-4b98-8730-2a80beee94d7.png)
![image](https://user-images.githubusercontent.com/94002623/236318394-6b3b5bdd-85bb-4792-a425-8a71d9e14cb1.png)
![image](https://user-images.githubusercontent.com/94002623/236424012-212e516a-5740-4eb0-91e6-25e0b500d726.png)

It also make difficult to QA (maybe a solution could be the one mentioned in #7) the 3d mapping because you don't understand if you messed up some min_heights/heights of it's the sloped terrain fault.

**Comment by sekerob** [comment]
There's multiple topics that seem strongly related, such as https://github.com/StrandedKitty/streets-gl/issues/35 and a comment elsewhere of 'eaten buildings' on hilly terrain. The second pic in issue 35 shows that phenomena at left. The mapped industrial area was actually levelled. Why the small substation floats is a true puzzle. But the building discussed with the missing building parts shows same, but it's on flat terrain. Elements on layer 2-3 such as the penthouse are floating as well. The heights of the parts definitely tie.

**Comment by ivanbranco** [comment]
I asked in #59 how type=building relations are used. I wonder if "gluing" all the members of a type=building relation is doable, that would fix this problem maybe.

**Comment by StrandedKitty** [comment]
I've added a toggle to disable terrain height. It should help with validation, but obviously most users aren't going to disable it, so they will observe broken detailed mapping.

Probably, it is possible to combine building parts into a single entity and apply elevation offset to this whole entity at once. It's straightforward for `type=building` relations, but for 3D buildings that don't use relations it might be tricky.

**Comment by Kovoschiz** [comment]
Yes, these cases that I suspected are why I suggested that. `type=building` is not mandatory, so `building=` is what's supposedly used for base elevation for the whole building I believe.

**Comment by ivanbranco** [comment]
> I've added a toggle to disable terrain height. It should help with validation, but obviously most users aren't going to disable it, so they will observe broken detailed mapping.

Yep, I was able to find some missing min_height, now is perfect 🐱 

![image](https://user-images.githubusercontent.com/94002623/236505473-3943ad8b-7703-4c5b-bc7e-a45cc9b33f1a.png)

Thank you!

**Comment by sekerob** [comment]
OK, excellent, hit Ctrl F5 to get the settings updated and disabled terrain elevation and even for flat terrain building complexes the hovering of parts has disappaered, now it looks truly as mapped.

![image](https://user-images.githubusercontent.com/23308504/236507663-6c210b4e-50bd-47f3-b827-ea5124fb9068.png)

Still have to figure out what's up with the roofgarden..

**Comment by janusbenissa** [comment]
![Screenshot 2023-05-21 at 16-23-44 Streets GL](https://github.com/StrandedKitty/streets-gl/assets/15191462/c57747be-0009-463b-9890-fa305b667961)

On a flat terrain it looks very good but in the hills :-(
Hopefully you find a solution.

**Comment by sanketgarade** [comment]
I found similar problem for this fort [wall](https://www.openstreetmap.org/way/105892152#map=18/18.83124/73.39507) on a mountain. The render makes it look as tall as the mountain in some places and non existent in some places.

Also to note that the [other wall](https://www.openstreetmap.org/way/1176968154) close to it, and having the same tags (`barrier=wall` and `building=yes`) renders properly.

[street gl link](https://streets.gl/#18.83167,73.39383,32.25,138.50,800.03) for quick reference.

![image](https://github.com/StrandedKitty/streets-gl/assets/74632379/a492fd95-e45d-471e-85a7-1bd80e1d481e)

![image](https://github.com/StrandedKitty/streets-gl/assets/74632379/c67184c9-adfa-4f56-bc7c-851da3a6e8a0)

**Comment by Mahabarata** [comment]
There is a problem with slopes and building:parts when all tags are the same but roof:direction (rendering is ok with terrain elevation disabled)

![2024-04-20_152053](https://github.com/StrandedKitty/streets-gl/assets/3121119/1bb378ee-04f0-45b9-9e5d-5dd5842747ee)

https://streets.gl/#45.71968,6.07378,24.75,331.75,55.74

The ways of the 2 offset roofs : https://www.openstreetmap.org/way/1061501830 and https://www.openstreetmap.org/way/1061501829
